README.md
lcd160x_driver
==============

A small linux kernel driver for LCD1602 and similiar displays connected via i2c. These displays are often used in arduino projects (https://arduino-info.wikispaces.com/LCD-Blue-I2C), so there really cheap and well known.

The driver is a character driver with write access. To clear the display (and run other functions of the hdd44780) ioctl's are implemented. Also a simle example on using it is included (see also https://github.com/tjohann/libbaalue.git).

WARNING: This is work in progress! Don't expect things to be complete in any dimension.


Requirement
-----------

You need at least gcc and kernel header to build the kernel driver and the hardware.

Used hardware: Bananapi-M1
Used defaults: I2C ...

![Alt text](pics/....png?raw=true "...")


Schematic
---------

I use ready to use modules, so there's no real schematic, but a simple description ([Howto connect](schematics/connecting_lcd1602.md)).


Usage
-----

The simple user interface to this repository is a Makefile. So type

    make

to get more info.


(Un)Installation
----------------

The driver creates a dev node pcf8574_lcd160x below /dev. A udev rule (udev/99-lcd160x.rules) is available:

	KERNEL=="lcd160x", MODE="0666"


Build the module:

	make build

This will build the driver and the example program ()


Install all:

	make install

This will install the driver to /lib/modules/$(shell uname -r)/kernel, install the udev rule 99-lcd160x.rules to /etc/udev/rules.d and the example program usage to ${HOME}/bin/usage_lcd160x


To get rid of them use:

	make uninstall

This will remove all installed parts.

Summarize:

![Alt text](pics/driver_build.png?raw=true "Build kernel module")


Load driver
-----------

Load the diver

    sudo modprobe lcd160x.ko

Depending on your distribution you can see info message from the driver in /var/log/messages.

	(sudo) tail -n 50 -f /var/log/messages

Summarize:

![Alt text](pics/driver_load.png?raw=true "Load the kernel module")


Use driver
----------

The userspace example (usage_lcd160x) will be installed in $(HOME)/bin. If you have it in your $PATH, then you can simply tape

    usage_lcd160x  .....

and see in /var/log/messages what the driver is doing.

Summarize:

![Alt text](pics/driver_usage.png?raw=true "Usage of driver")


Documentation
-------------

Below the folder Documentation you can find information about the hardware and more.


Pictures/Test-environment
-------------------------

Here you find some pictures of the wiring and my test setup.

![Alt text](pics/overview.jpg?raw=true "Overview ...")


NEWS
#
# NEWS
#

2017-11-19
- fix build errors (asm/uaccess.h -> linux/uaccess.h)

2017-02-17
- rename driver and usage to (*)pcf8574_lcd160x
- some first steps toward of integration
- first steps to make the driver aware of LCD160x and LCD200x

2016-09-26
- some cleanups

2016-09-14
- initial skeleton based on my_driver.git repository

common.h
/*
 * lcd160x.c -> driver for lcd1602 displays
 *
 * GPL
 * (c) 2013-2017, thorsten.johannvorderbrueggen@t-online.de
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifndef _COMMON_H_
#define _COMMON_H_

#define DRIVER_NAME "pcf8574_lcd160x"
#define DEV_NAME "/dev/pcf8574_lcd160x"

#define IOCTL_SET_I2C_CLIENT 0x0001

struct client_config {
	unsigned int pin_irq;
	int adapter_nr;
	unsigned short addr;
};

#endif


usage/usage.c
/*
 * usage.c -> show usage of pcf8574_lcd160x driver
 *
 * GPL
 * (c) 2016-2017, thorsten.johannvorderbrueggen@t-online.de
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <sys/ioctl.h>
#include <stdbool.h>
#include <time.h>

/* common defines for driver and usage */
#include "../common.h"

#define MAX_LINE 256

static void
__attribute__((noreturn)) usage(void)
{
	fprintf(stdout, "Usage: ./usage -[p PIN] -b adapter -a addr \n");
	fprintf(stdout, "       -p -> PIN for IRQ                   \n");
	fprintf(stdout, "       -b -> Adapter number                \n");
	fprintf(stdout, "       -a -> Address (hex)                 \n");
	putchar('\n');
	fprintf(stdout, "Examples:                                  \n");
	fprintf(stdout, "       ./usage -a 0x27 -b 2                \n");
	exit(EXIT_FAILURE);
}

static void
print_client_config(struct client_config *client)
{
	printf("PIN for IRQ:   %d\n", client->pin_irq);
	printf("Adapternumber: %d\n", client->adapter_nr);
	printf("Address:       0x%x\n", client->addr);
}

static bool
is_client_config_valid(struct client_config *client)
{
	if (client->pin_irq == 0)
		printf("wont use IRQ\n");

	if (client->adapter_nr > 20)
		printf("adapter number > %d makes no sense\n",
		       client->adapter_nr);

	if (client->addr > 128)
		printf("adapter number > %d makes no sense\n",
		       client->addr);
	return true;
}

static int
work_mode(int fd)
{
	struct timespec t;

	memset(&t, 0, sizeof(struct timespec));

	/* sleep time 0.5 sec */
	t.tv_sec = 0;
	t.tv_nsec = 500000000;

	int value = 0;
	size_t len = sizeof(value);
	ssize_t n = 0;

	/* toogle port bits */
	for (;;) {
		n = write(fd, &value, len);
		if (n == -1)
			perror("write");
		clock_nanosleep(CLOCK_MONOTONIC, 0, &t, NULL);

		value = 1;
		n = write(fd, &value, len);
		if (n == -1)
			perror("write");
		clock_nanosleep(CLOCK_MONOTONIC, 0, &t, NULL);
		value = 0;
	}

	return 0; /* should never reached */
}

static int
open_device(void)
{
	int fd = -1;

	fd = open(DEV_NAME, O_WRONLY);
	if (fd == -1) {
		fprintf(stderr, "could not open %s\n", DEV_NAME);
		return -1;
	}
	return fd;
}

int
main(int argc, char *argv[])
{
	struct client_config client;
	memset(&client, 0, sizeof(client));

	int c;
	while ((c = getopt(argc, argv, "p:b:a:h")) != -1) {
		switch (c) {
		case 'p':
			client.pin_irq = atoi(optarg);
			break;
		case 'b':
			client.adapter_nr = atoi(optarg);
			break;
		case 'a':
			client.addr = strtoul(optarg, NULL, 16);
			break;
		case 'h':
			usage();
			break;
		default:
			fprintf(stderr, "ERROR: no valid argument\n");
			usage();
		}
	}

	print_client_config(&client);

	if (!is_client_config_valid(&client)) {
		fprintf(stderr,"client config is not valid\n");
		usage();
	}

	int fd = open_device();
	if (fd == -1)
		usage();

	int ret = ioctl(fd, IOCTL_SET_I2C_CLIENT , &client);
	if (ret == -1) {
		fprintf(stderr, "could not set client values\n");
		return EXIT_FAILURE;
	}

	if (work_mode(fd) == -1)
		usage();

	close(fd);
	return EXIT_SUCCESS;
}


driver/lcd160x.c
/*
 * lcd160x.c -> driver for lcd160x display
 *
 * GPL
 * (c) 2016-2017, thorsten.johannvorderbrueggen@t-online.de
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <linux/module.h>
#include <linux/fs.h>
#include <linux/cdev.h>
#include <linux/device.h>
#include <asm/uaccess.h>
#include <linux/i2c.h>
#include <linux/slab.h>

#include "../common.h"

static int pcf8574_probe(struct i2c_client *client,
			 const struct i2c_device_id *id);
static int pcf8574_remove(struct i2c_client *client);

static dev_t pcf8574_dev_number;
static struct cdev *pcf8574_object;
struct class *pcf8574_class;
static struct device *drv_dev;

struct _instance_data {
	struct i2c_client *slave;
	struct i2c_adapter *adapter;
	unsigned int pin_irq;
	int gpio_irq;
	int adapter_nr;
	unsigned short addr;
};
#define SD struct _instance_data

static struct i2c_device_id pcf8574_idtable[] = {
        { "pcf8574_lcd160x", 0 },
	{ }
};
MODULE_DEVICE_TABLE(i2c, pcf8574_idtable);

static struct i2c_driver pcf8574_driver = {
        .driver = {
                .name   = "pcf8574_lcd160x",
        },
        .id_table       = pcf8574_idtable,
        .probe          = pcf8574_probe,
        .remove         = pcf8574_remove,
};

static ssize_t
lcd160x_write(struct file *instance,
	      const char __user *user, size_t count, loff_t *offset)
{
	int not_send;
	unsigned char value;

	SD *data = NULL;

	if (!instance->private_data) {
		dev_err(drv_dev, "write: instance->private_data == NULL\n");
		return -EFAULT;
	}

	data = (SD *) instance->private_data;
	if (!data->slave) {
		dev_err(drv_dev, "no valid slave pointer\n");
		return -EFAULT;
	}

	if (get_user(value, (unsigned char __user *) user)) {
		dev_err(drv_dev, "could not copy from userspace\n");
		return -EFAULT;
	}

	dev_info(drv_dev, "got %d from userspace\n", value);

	not_send = i2c_master_send(data->slave, &value, sizeof(value));
	if (not_send < 0) {
		dev_err(drv_dev, "i2c_master_send\n");
		return -EFAULT;
	}

	return 0;
}

static int
lcd160x_open(struct inode *dev_node, struct file *instance)
{
	SD *data = NULL;

	int accmode = (instance->f_flags & O_ACCMODE);
	bool write_mode = (accmode == O_WRONLY);

	if (!write_mode) {
		dev_err(drv_dev, "only O_WRONLY allowed\n");
		return -EIO;
	}

	data = (SD *) kmalloc(sizeof(SD), GFP_USER);
	if (data == NULL) {
		dev_err(drv_dev, "kmalloc\n");
		return -EIO;
	}
	memset(data, 0, sizeof(SD));

	instance->private_data = (void *) data;

	return 0;
}

static int
lcd160x_close(struct inode *dev_node, struct file *instance)
{
	SD *data = NULL;

	if (instance->private_data) {
		data = (SD *) instance->private_data;

		if (data->slave)
			i2c_unregister_device(data->slave);

		kfree(instance->private_data);
	} else {
		dev_err(drv_dev, "close: instance->private_data == NULL\n");
		return -EFAULT;
	}

	return 0;
}

static long
lcd160x_ioctl(struct file *instance, unsigned int cmd,
		  unsigned long __user arg)
{
	struct client_config __user user_data;
	SD *data = NULL;

	if (!instance->private_data) {
		dev_err(drv_dev, "ioctl: instance->private_data == NULL\n");
		return -1;
	}

	data = (SD *) instance->private_data;

        /*
	 * TODO: cmd needed
	 *       clear display
	 *       clear buffer (which rotates?)
	 *       set cursor
	 *       set buffer size to rotate?
	 *       set size of display (16x2 or 16x4 or ...)
	 */
	switch (cmd) {
	case IOCTL_SET_I2C_CLIENT:
		if (copy_from_user(&user_data,
				   (struct client_config __user *) arg,
				   sizeof(struct client_config)))
			return -EFAULT;

		dev_info(drv_dev, "addr: 0x%x\n", user_data.addr);
		dev_info(drv_dev, "adapter_nr: %d\n", user_data.adapter_nr);
		dev_info(drv_dev, "IRQ - pin %d\n", user_data.pin_irq);

		struct i2c_board_info info = {
			I2C_BOARD_INFO("pcf8574_lcd160x", user_data.addr),
		};

		data->adapter = i2c_get_adapter(user_data.adapter_nr);
		if (data->adapter == NULL)
			goto error;

		data->slave = i2c_new_device(data->adapter, &info);
		if (data->slave == NULL)
			goto error;

		data->addr = user_data.addr;
		data->adapter_nr = user_data.adapter_nr;

		break;
	default:
		dev_err(drv_dev, "unknown ioctl 0x%x\n", cmd);
		return -EINVAL;
	}

	return 0;
error:
	dev_err(drv_dev, "i2c_new_device or i2c_get_adapter\n");
	i2c_del_driver(&pcf8574_driver);

	return -EFAULT;
}

static struct file_operations fops = {
	.owner= THIS_MODULE,
	.write= lcd160x_write,
	.unlocked_ioctl = lcd160x_ioctl,
	.open = lcd160x_open,
	.release = lcd160x_close
};

static int
pcf8574_probe(struct i2c_client *client,
	      const struct i2c_device_id *id)
{
	dev_info(drv_dev, "client %p\n", client);
	dev_info(drv_dev, "client->addr 0x%x", client->addr);
	dev_info(drv_dev, "id %p\n", id);
	dev_info(drv_dev, "id->name %s\n", id->name);

	/*
	 * TODO: clear the display and display a text like
	 * "driver %s is ready", driver_name ... or so
	 */

	return 0;
}

static int
pcf8574_remove(struct i2c_client *client)
{
	dev_info(drv_dev, "pcf8574_remove\n");

	/*
	 * TODO: clear display
	 */

	return 0;
}

static int __init
i2c_lcd160x_init(void)
{
	pr_info("i2c_lcd160x_init called\n");

	if (alloc_chrdev_region(&pcf8574_dev_number, 0, 1, DRIVER_NAME) < 0)
		return -EIO;

	pcf8574_object = cdev_alloc();
	if (pcf8574_object == NULL) {
		dev_err(drv_dev, "cdev_alloc\n");
		goto free_dev_number;
	}

	pcf8574_object->owner = THIS_MODULE;
	pcf8574_object->ops = &fops;

	if (cdev_add(pcf8574_object, pcf8574_dev_number, 1)) {
		dev_err(drv_dev, "cdev_add\n");
		goto free_cdev;
	}

	pcf8574_class = class_create(THIS_MODULE, DRIVER_NAME);
	if (IS_ERR(pcf8574_class)) {
		dev_err(drv_dev, "class_create\n");
		goto free_cdev;
	}

	drv_dev = device_create(pcf8574_class, NULL, pcf8574_dev_number,
				NULL, "%s", DRIVER_NAME);
	if (IS_ERR(drv_dev)) {
		dev_err(drv_dev, "device_create\n");
		goto free_class;
	}

	if (i2c_add_driver(&pcf8574_driver)) {
		dev_err(drv_dev, "i2c_add_driver\n");
		goto free_class;
	}

	return 0;

free_class:
	device_destroy(pcf8574_class, pcf8574_dev_number);
	class_destroy(pcf8574_class );

free_cdev:
	kobject_put(&pcf8574_object->kobj);

free_dev_number:
	unregister_chrdev_region(pcf8574_dev_number, 1);

	return -EIO;
}

static void __exit
i2c_lcd160x_exit(void)
{
	device_destroy(pcf8574_class, pcf8574_dev_number);
	class_destroy(pcf8574_class );

	cdev_del(pcf8574_object);
	unregister_chrdev_region(pcf8574_dev_number, 1);

	i2c_del_driver(&pcf8574_driver);

	return;
}

module_init(i2c_lcd160x_init);
module_exit(i2c_lcd160x_exit);

MODULE_LICENSE("GPL");
MODULE_DESCRIPTION("pcf8574 base lcd1602 driver");
MODULE_AUTHOR("Thorsten Johannvorderbrueggen <thorsten.johannvorderbrueggen@t-online.de>");


driver/pcf8574_lcd160x.c
/*
 * lcd160x.c -> driver for lcd160x display
 *
 * GPL
 * (c) 2016-2017, thorsten.johannvorderbrueggen@t-online.de
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <linux/module.h>
#include <linux/fs.h>
#include <linux/cdev.h>
#include <linux/device.h>
#include <linux/uaccess.h>
#include <linux/i2c.h>
#include <linux/slab.h>

#include "../common.h"

static int pcf8574_probe(struct i2c_client *client,
			 const struct i2c_device_id *id);
static int pcf8574_remove(struct i2c_client *client);

static long lcd160x_ioctl(struct file *instance, unsigned int cmd,
			  unsigned long __user arg);
static ssize_t lcd160x_write(struct file *instance,
			     const char __user *user,
			     size_t count, loff_t *offset);
static int lcd160x_open(struct inode *dev_node, struct file *instance);
static int lcd160x_close(struct inode *dev_node, struct file *instance);

static dev_t pcf8574_dev_number;
static struct cdev *pcf8574_object;
struct class *pcf8574_class;
static struct device *drv_dev;

struct _instance_data {
	struct i2c_client *slave;
	struct i2c_adapter *adapter;
	unsigned int pin_irq;
	int gpio_irq;
	int adapter_nr;
	unsigned short addr;
	unsigned char num_lines;
	unsigned char num_rows;
};
#define SD struct _instance_data

static struct i2c_device_id pcf8574_idtable[] = {
        { "pcf8574_lcd160x", 0 },
	{ "pcf8574_lcd200x", 1 },
	{ }
};
MODULE_DEVICE_TABLE(i2c, pcf8574_idtable);

#ifdef CONFIG_OF
static const struct of_device_id pcf8574_of_table[] = {
	{ .compatible = "nxp,pcf8574_lcd160x" },
	{ .compatible = "nxp,pcf8574_lcd200x" },
	{ }
};
MODULE_DEVICE_TABLE(of, pcf8574_of_table);
#endif

static struct i2c_driver pcf8574_driver = {
        .driver = {
                .name   = "pcf8574_lcd160x",
		.of_match_table = of_match_ptr(pcf8574_of_table),
        },
        .id_table       = pcf8574_idtable,
        .probe          = pcf8574_probe,
        .remove         = pcf8574_remove,
};

static struct file_operations fops = {
       .owner= THIS_MODULE,
       .write= lcd160x_write,
       .unlocked_ioctl = lcd160x_ioctl,
       .open = lcd160x_open,
       .release = lcd160x_close
};


/*-------------------------------------------------------------------------*/

static ssize_t
lcd160x_write(struct file *instance,
	      const char __user *user, size_t count, loff_t *offset)
{
	int not_send;
	unsigned char value;

	SD *data = NULL;

	if (!instance->private_data) {
		dev_err(drv_dev, "write: instance->private_data == NULL\n");
		return -EFAULT;
	}

	data = (SD *) instance->private_data;
	if (!data->slave) {
		dev_err(drv_dev, "no valid slave pointer\n");
		return -EFAULT;
	}

	if (get_user(value, (unsigned char __user *) user)) {
		dev_err(drv_dev, "could not copy from userspace\n");
		return -EFAULT;
	}

	dev_info(drv_dev, "got %d from userspace\n", value);

	not_send = i2c_master_send(data->slave, &value, sizeof(value));
	if (not_send < 0) {
		dev_err(drv_dev, "i2c_master_send\n");
		return -EFAULT;
	}

	return 0;
}

static int
lcd160x_open(struct inode *dev_node, struct file *instance)
{
	SD *data = NULL;

	int accmode = (instance->f_flags & O_ACCMODE);
	bool write_mode = (accmode == O_WRONLY);

	if (!write_mode) {
		dev_err(drv_dev, "only O_WRONLY allowed\n");
		return -EIO;
	}

	data = (SD *) kmalloc(sizeof(SD), GFP_USER);
	if (data == NULL) {
		dev_err(drv_dev, "kmalloc\n");
		return -EIO;
	}
	memset(data, 0, sizeof(SD));

	instance->private_data = (void *) data;

	return 0;
}

static int
lcd160x_close(struct inode *dev_node, struct file *instance)
{
	SD *data = NULL;

	if (instance->private_data) {
		data = (SD *) instance->private_data;

		if (data->slave)
			i2c_unregister_device(data->slave);

		kfree(instance->private_data);
	} else {
		dev_err(drv_dev, "close: instance->private_data == NULL\n");
		return -EFAULT;
	}

	return 0;
}

static long
lcd160x_ioctl(struct file *instance, unsigned int cmd,
		  unsigned long __user arg)
{
	struct client_config __user user_data;
	SD *data = NULL;

	if (!instance->private_data) {
		dev_err(drv_dev, "ioctl: instance->private_data == NULL\n");
		return -1;
	}

	data = (SD *) instance->private_data;

        /*
	 * TODO: cmd needed
	 *       clear display
	 *       clear buffer (which rotates?)
	 *       set cursor
	 *       set buffer size to rotate?
	 *       set size of display (16x2 or 16x4 or ...)
	 */
	switch (cmd) {
	case IOCTL_SET_I2C_CLIENT:
		if (copy_from_user(&user_data,
				   (struct client_config __user *) arg,
				   sizeof(struct client_config)))
			return -EFAULT;

		dev_info(drv_dev, "addr: 0x%x\n", user_data.addr);
		dev_info(drv_dev, "adapter_nr: %d\n", user_data.adapter_nr);
		dev_info(drv_dev, "IRQ - pin %d\n", user_data.pin_irq);

		struct i2c_board_info info = {
			I2C_BOARD_INFO("pcf8574_lcd160x", user_data.addr),
		};

		data->adapter = i2c_get_adapter(user_data.adapter_nr);
		if (data->adapter == NULL)
			goto error;

		data->slave = i2c_new_device(data->adapter, &info);
		if (data->slave == NULL)
			goto error;

		data->addr = user_data.addr;
		data->adapter_nr = user_data.adapter_nr;

		break;
	default:
		dev_err(drv_dev, "unknown ioctl 0x%x\n", cmd);
		return -EINVAL;
	}

	return 0;
error:
	dev_err(drv_dev, "i2c_new_device or i2c_get_adapter\n");
	i2c_del_driver(&pcf8574_driver);

	return -EFAULT;
}

/*-------------------------------------------------------------------------*/

static int
pcf8574_probe(struct i2c_client *client,
	      const struct i2c_device_id *id)
{
	dev_info(drv_dev, "client %p\n", client);
	dev_info(drv_dev, "client->addr 0x%x", client->addr);
	dev_info(drv_dev, "id %p\n", id);
	dev_info(drv_dev, "id->name %s\n", id->name);
	dev_info(drv_dev, "id->driver_data %lu\n", id->driver_data);

	/*
	 * TODO: check driver_data to distinguish between LCD1602 and LCD2004
	 */

	/*
	 * TODO: clear the display and display a text like
	 * "driver %s is ready", driver_name ... or so
	 */

	return 0;
}

static int
pcf8574_remove(struct i2c_client *client)
{
	dev_info(drv_dev, "pcf8574_remove\n");

	/*
	 * TODO: clear display
	 */

	return 0;
}

/*-------------------------------------------------------------------------*/

static int __init
i2c_lcd160x_init(void)
{
	pr_info("i2c_lcd160x_init called\n");

	if (alloc_chrdev_region(&pcf8574_dev_number, 0, 1, DRIVER_NAME) < 0)
		return -EIO;

	pcf8574_object = cdev_alloc();
	if (pcf8574_object == NULL) {
		dev_err(drv_dev, "cdev_alloc\n");
		goto free_dev_number;
	}

	pcf8574_object->owner = THIS_MODULE;
	pcf8574_object->ops = &fops;

	if (cdev_add(pcf8574_object, pcf8574_dev_number, 1)) {
		dev_err(drv_dev, "cdev_add\n");
		goto free_cdev;
	}

	pcf8574_class = class_create(THIS_MODULE, DRIVER_NAME);
	if (IS_ERR(pcf8574_class)) {
		dev_err(drv_dev, "class_create\n");
		goto free_cdev;
	}

	drv_dev = device_create(pcf8574_class, NULL, pcf8574_dev_number,
				NULL, "%s", DRIVER_NAME);
	if (IS_ERR(drv_dev)) {
		dev_err(drv_dev, "device_create\n");
		goto free_class;
	}

	if (i2c_add_driver(&pcf8574_driver)) {
		dev_err(drv_dev, "i2c_add_driver\n");
		goto free_class;
	}

	return 0;

free_class:
	device_destroy(pcf8574_class, pcf8574_dev_number);
	class_destroy(pcf8574_class );

free_cdev:
	kobject_put(&pcf8574_object->kobj);

free_dev_number:
	unregister_chrdev_region(pcf8574_dev_number, 1);

	return -EIO;
}

static void __exit
i2c_lcd160x_exit(void)
{
	device_destroy(pcf8574_class, pcf8574_dev_number);
	class_destroy(pcf8574_class );

	cdev_del(pcf8574_object);
	unregister_chrdev_region(pcf8574_dev_number, 1);

	i2c_del_driver(&pcf8574_driver);

	return;
}

module_init(i2c_lcd160x_init);
module_exit(i2c_lcd160x_exit);

MODULE_LICENSE("GPL");
MODULE_DESCRIPTION("pcf8574 base lcd1602 driver");
MODULE_AUTHOR("Thorsten Johannvorderbrueggen <thorsten.johannvorderbrueggen@t-online.de>");
